File: kernel_design/executable_spec/kernel/task_runtime.py
from __future__ import annotations

from abc import ABC, abstractmethod
from collections.abc import Mapping as MappingABC
from dataclasses import dataclass
from enum import Enum
from typing import ClassVar, Mapping, Sequence

from .task_score_band import Suffix, TaskId, TimeMillis
from .worker_runtime import EventCode, WorkerGroupId


MessageId = str
ItemPriority = int


class TaskType(Enum):
    TASK_DRIVEN = "TASK_DRIVEN"
    ITEM_DRIVEN = "ITEM_DRIVEN"


@dataclass(frozen=True, slots=True, kw_only=True)
class TaskItem:
    """Canonical Task-scoped item record; scheduling state lives elsewhere."""

    message_id: MessageId
    event_code: EventCode
    created_at_millis: TimeMillis
    payload: Mapping[str, object]
    priority: ItemPriority = 5
    expire_at_millis: TimeMillis | None = None
    allocation_rule: Mapping[str, object] | None = None

    def __post_init__(self) -> None:
        if not self.message_id:
            raise ValueError("message id must be non-empty")
        if not self.event_code:
            raise ValueError("event code must be non-empty")
        if not isinstance(self.payload, MappingABC):
            raise ValueError("task item payload must be a mapping")
        if self.allocation_rule is not None and (
            not isinstance(self.allocation_rule, MappingABC)
            or not self.allocation_rule
        ):
            raise ValueError("task item allocation rule must be a non-empty mapping")
        if (
            not isinstance(self.priority, int)
            or isinstance(self.priority, bool)
            or not 0 <= self.priority <= 10
        ):
            raise ValueError("task item priority must be in 0..10")
        if self.created_at_millis < 0:
            raise ValueError("created_at_millis must be non-negative")
        if (
            self.expire_at_millis is not None
            and self.expire_at_millis <= self.created_at_millis
        ):
            raise ValueError("expire_at_millis must be after created_at_millis")


class TaskItemAppendStatus(Enum):
    APPENDED = "appended"
    RETRYABLE = "retryable"
    NOT_FOUND = "not_found"
    INVALID = "invalid"


@dataclass(frozen=True)
class TaskItemAppendResult:
    status: TaskItemAppendStatus
    reason: str | None = None


@dataclass(frozen=True)
class TaskDescriptor:
    """Stable Task scheduling metadata, separate from score runtime truth."""

    task_id: TaskId
    worker_group_id: WorkerGroupId
    task_type: TaskType
    allocation_rule: Mapping[str, object] | None
    config: Mapping[str, str]
    empty_close_at_millis: TimeMillis | None = None

    CONFIG_KEYS: ClassVar[frozenset[str]] = frozenset(
        {
            "priority",
            "maximumCandidateWorkers",
            "maxRetryTimes",
        }
    )

    def __post_init__(self) -> None:
        if not self.task_id:
            raise ValueError("task id must be non-empty")
        if not self.worker_group_id:
            raise ValueError("worker group id must be non-empty")
        if not isinstance(self.task_type, TaskType):
            raise ValueError("task type is invalid")
        if self.task_type is TaskType.TASK_DRIVEN:
            if not isinstance(self.allocation_rule, MappingABC):
                raise ValueError("TASK_DRIVEN requires a Task allocation rule")
        elif self.allocation_rule is not None:
            raise ValueError("ITEM_DRIVEN forbids a Task allocation rule")
        if not isinstance(self.config, MappingABC):
            raise ValueError("task config must be a mapping")
        if self.empty_close_at_millis is not None and (
            not isinstance(self.empty_close_at_millis, int)
            or isinstance(self.empty_close_at_millis, bool)
            or self.empty_close_at_millis < 0
        ):
            raise ValueError("empty_close_at_millis must be non-negative")
        if set(self.config) != self.CONFIG_KEYS:
            raise ValueError("task config must contain exactly the declared keys")
        if any(
            not isinstance(value, str)
            for value in self.config.values()
        ):
            raise ValueError("task config values must be strings")

        priority = self._decimal_config("priority")
        maximum_candidates = self._decimal_config("maximumCandidateWorkers")
        max_retry_times = self._decimal_config("maxRetryTimes")
        if not 0 <= priority <= 99:
            raise ValueError("task priority must be in 0..99")
        if maximum_candidates <= 0:
            raise ValueError("maximum candidate workers must be positive")
        if not 0 <= max_retry_times <= 98:
            raise ValueError("max retry times must be in 0..98")

    def _decimal_config(self, key: str) -> int:
        value = self.config[key]
        if not value.isascii() or not value.isdecimal():
            raise ValueError(f"task config {key} must be decimal text")
        return int(value)


class TaskCreationStatus(Enum):
    CREATED = "created"
    RETRYABLE = "retryable"
    CONFLICT = "conflict"
    INVALID = "invalid"


@dataclass(frozen=True)
class TaskCreationResult:
    status: TaskCreationStatus
    reason: str | None = None


class TaskRuntime(ABC):
    """Task, canonical TaskItem, and last-success result owner surface."""

    @abstractmethod
    def create_task(
        self,
        *,
        descriptor: TaskDescriptor,
        suffix: Suffix,
    ) -> TaskCreationResult:
        """Create one Task from a descriptor with resolved empty-close time."""
        pass

    @abstractmethod
    def append_items(
        self,
        *,
        task_id: TaskId,
        items: Sequence[TaskItem],
    ) -> Mapping[MessageId, TaskItemAppendResult]:
        """Store the latest records and initialize missing Item scores."""
        pass

    @abstractmethod
    def load_task_items(
        self,
        *,
        task_id: TaskId,
        message_ids: Sequence[MessageId],
    ) -> Mapping[MessageId, TaskItem | None]:
        """Load canonical records for one bounded Task-scoped message-id batch."""
        pass

    @abstractmethod
    def store_task_item_success_results(
        self,
        *,
        task_id: TaskId,
        results: Mapping[MessageId, str],
    ) -> None:
        """Store the latest successful payload for each Task-scoped Item."""
        pass

    @abstractmethod
    def load_task_item_success_results(
        self,
        *,
        task_id: TaskId,
        message_ids: Sequence[MessageId],
    ) -> Mapping[MessageId, str | None]:
        """Load bounded Task-scoped last-success result payloads."""
        pass


class TaskResourceCatalog(ABC):
    """Bounded allocation descriptor reads.

    This surface does not own task score, lifecycle, work, allocation handoff,
    result state, or query/list APIs. Task ids are globally unique.
    """

    @abstractmethod
    def load_task_allocation_descriptors(
        self,
        *,
        task_ids: Sequence[TaskId],
    ) -> Mapping[TaskId, TaskDescriptor | None]:
        """Load one bounded allocation batch; missing ids map to None.

        This is an assignment-dispatch input surface, not a general Task read
        or query API.
        """
        pass

File: kernel_jvm/src/main/java/com/xa/mass/kernel/task/TaskRuntime.java
package com.xa.mass.kernel.task;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import org.jspecify.annotations.Nullable;

public interface TaskRuntime {

    int MAX_RESULT_SCAN_COUNT_HINT = 1000;

    Set<String> CONFIG_KEYS = Set.of(
            "priority",
            "maximumCandidateWorkers",
            "maxRetryTimes"
    );

    TaskCreationResult createTask(TaskDescriptor descriptor);

    Map<String, TaskItemAppendResult> appendItems(
            String taskId,
            List<TaskItem> items
    );

    Map<String, @Nullable TaskItem> loadTaskItems(
            String taskId,
            List<String> messageIds
    );

    void storeTaskItemSuccessResults(
            String taskId,
            Map<String, String> results
    );

    void storeTaskItemFailedResults(
            String taskId,
            List<String> messageIds
    );

    Map<String, @Nullable TaskItemResult> loadTaskItemResults(
            String taskId,
            List<String> messageIds
    );

    TaskItemResultPage scanTaskItemResults(
            String taskId,
            String cursor,
            int countHint
    );

    enum WorkerAllocationMechanism {
        PRECOMPUTED_TASK_RULE,
        ON_DEMAND_ITEM_RULE
    }

    enum TaskIdleDisposition {
        CLOSE_WHEN_IDLE,
        PARK_WHEN_IDLE
    }

    enum TaskItemAppendStatus {
        APPENDED("appended"),
        RETRYABLE("retryable"),
        NOT_FOUND("not_found"),
        INVALID("invalid");

        private final String wireValue;

        TaskItemAppendStatus(String wireValue) {
            this.wireValue = wireValue;
        }

        public String wireValue() {
            return wireValue;
        }
    }

    enum TaskCreationStatus {
        CREATED("created"),
        RETRYABLE("retryable"),
        CONFLICT("conflict"),
        INVALID("invalid");

        private final String wireValue;

        TaskCreationStatus(String wireValue) {
            this.wireValue = wireValue;
        }

        public String wireValue() {
            return wireValue;
        }
    }

    record TaskItem(
            String messageId,
            String eventCode,
            long createdAtMillis,
            Map<String, Object> payload,
            int priority,
            @Nullable Long expireAtMillis,
            @Nullable Map<String, Object> allocationRule
    ) {
        public TaskItem {
            requireNonBlank(messageId, "messageId");
            requireNonBlank(eventCode, "eventCode");
            Objects.requireNonNull(payload, "payload");
            if (priority < 0 || priority > 10) {
                throw new IllegalArgumentException(
                        "priority must be in 0..10"
                );
            }
            if (createdAtMillis < 0) {
                throw new IllegalArgumentException(
                        "createdAtMillis must be non-negative"
                );
            }
            if (expireAtMillis != null
                    && expireAtMillis <= createdAtMillis) {
                throw new IllegalArgumentException(
                        "expireAtMillis must be after createdAtMillis"
                );
            }
            payload = immutableObjectMap(payload);
            allocationRule = allocationRule == null
                    ? null
                    : immutableObjectMap(allocationRule);
        }
    }

    record TaskDescriptor(
            String taskId,
            String workerGroupId,
            WorkerAllocationMechanism workerAllocationMechanism,
            TaskIdleDisposition idleDisposition,
            @Nullable Map<String, Object> allocationRule,
            Map<String, String> config
    ) {
        public TaskDescriptor {
            requireNonBlank(taskId, "taskId");
            requireNonBlank(workerGroupId, "workerGroupId");
            Objects.requireNonNull(
                    workerAllocationMechanism,
                    "workerAllocationMechanism"
            );
            Objects.requireNonNull(idleDisposition, "idleDisposition");
            Objects.requireNonNull(config, "config");
            if (workerAllocationMechanism
                    == WorkerAllocationMechanism.PRECOMPUTED_TASK_RULE
                    && allocationRule == null) {
                throw new IllegalArgumentException(
                        "PRECOMPUTED_TASK_RULE requires allocationRule"
                );
            }
            if (workerAllocationMechanism
                    == WorkerAllocationMechanism.ON_DEMAND_ITEM_RULE
                    && allocationRule != null) {
                throw new IllegalArgumentException(
                        "ON_DEMAND_ITEM_RULE forbids allocationRule"
                );
            }
            if (!config.keySet().equals(CONFIG_KEYS)) {
                throw new IllegalArgumentException(
                        "config must contain exactly the declared keys"
                );
            }
            int priority = decimalConfig(config, "priority");
            int maximumCandidates = decimalConfig(
                    config,
                    "maximumCandidateWorkers"
            );
            int maxRetryTimes = decimalConfig(config, "maxRetryTimes");
            if (priority < 0 || priority > 99) {
                throw new IllegalArgumentException(
                        "Task priority must be in 0..99"
                );
            }
            if (maximumCandidates <= 0) {
                throw new IllegalArgumentException(
                        "maximumCandidateWorkers must be positive"
                );
            }
            if (maxRetryTimes < 0 || maxRetryTimes > 98) {
                throw new IllegalArgumentException(
                        "maxRetryTimes must be in 0..98"
                );
            }
            allocationRule = allocationRule == null
                    ? null
                    : immutableObjectMap(allocationRule);
            config = Collections.unmodifiableMap(
                    new LinkedHashMap<>(config)
            );
        }
    }

    record TaskItemAppendResult(
            TaskItemAppendStatus status,
            @Nullable String reason
    ) {
        public TaskItemAppendResult {
            Objects.requireNonNull(status, "status");
        }

        public TaskItemAppendResult(TaskItemAppendStatus status) {
            this(status, null);
        }
    }

    record TaskItemResult(
            String code,
            String opaqueResultPayload
    ) {

        private static final String SUCCESS_CODE = "200";
        private static final String FAILED_CODE = "failed";
        private static final String FAILED_PAYLOAD =
                "TaskItem ended without a successful result";

        public TaskItemResult {
            requireNonBlank(code, "code");
            requireNonBlank(
                    opaqueResultPayload,
                    "opaqueResultPayload"
            );
        }

        public boolean succeeded() {
            return SUCCESS_CODE.equals(code);
        }

        public static TaskItemResult succeeded(String payload) {
            return new TaskItemResult(SUCCESS_CODE, payload);
        }

        public static TaskItemResult failed() {
            return new TaskItemResult(FAILED_CODE, FAILED_PAYLOAD);
        }
    }

    record TaskItemResultPage(
            String nextCursor,
            Map<String, TaskItemResult> results
    ) {
        public TaskItemResultPage {
            requireDecimal(nextCursor, "nextCursor");
            Objects.requireNonNull(results, "results");
            results.forEach((messageId, result) -> {
                requireNonBlank(messageId, "result messageId");
                Objects.requireNonNull(result, "result");
            });
            results = Collections.unmodifiableMap(
                    new LinkedHashMap<>(results)
            );
        }
    }

    record TaskCreationResult(
            TaskCreationStatus status,
            @Nullable String reason
    ) {
        public TaskCreationResult {
            Objects.requireNonNull(status, "status");
        }

        public TaskCreationResult(TaskCreationStatus status) {
            this(status, null);
        }
    }

    private static int decimalConfig(
            Map<String, String> config,
            String key
    ) {
        String value = config.get(key);
        if (value == null || value.isEmpty()) {
            throw new IllegalArgumentException(
                    "config " + key + " must be decimal text"
            );
        }
        for (int index = 0; index < value.length(); index++) {
            char character = value.charAt(index);
            if (character < '0' || character > '9') {
                throw new IllegalArgumentException(
                        "config " + key + " must be decimal text"
                );
            }
        }
        return Integer.parseInt(value);
    }

    private static Map<String, Object> immutableObjectMap(
            Map<String, Object> source
    ) {
        return Collections.unmodifiableMap(new LinkedHashMap<>(source));
    }

    private static void requireNonBlank(String value, String name) {
        if (value == null || value.isEmpty()) {
            throw new IllegalArgumentException(name + " must be non-empty");
        }
    }

    private static void requireDecimal(String value, String name) {
        requireNonBlank(value, name);
        for (int index = 0; index < value.length(); index++) {
            char character = value.charAt(index);
            if (character < '0' || character > '9') {
                throw new IllegalArgumentException(
                        name + " must be decimal text"
                );
            }
        }
    }
}
